### DISCUSSION TITLE
Guide for self hosting wakapi on AWS EC2

### DISCUSSION THREAD

**Discussion by palash-sarate** [muety/wakapi#651] [General]
[config.yml.txt](https://github.com/user-attachments/files/15894009/config.yml.txt)
[wakapi.service.txt](https://github.com/user-attachments/files/15894019/wakapi.service.txt)

attached is my attempt for self hosting the wakapi on an ec2 instance of AWS. i have created a service for wakapi but with attached config i get error 

```
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.80983168Z [INFO ] no need to migrate '20210806-remove_persisted_project_labels'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.809913283Z [INFO ] potentially running migration '20211215-migrate_id_to_bigint-add_has_data_field'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.8100554Z [INFO ] no need to migrate '20211215-migrate_id_to_bigint-add_has_data_field'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810132713Z [INFO ] potentially running migration '20212212-total_summary_heartbeats'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810282783Z [INFO ] no need to migrate '20212212-total_summary_heartbeats'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810363338Z [INFO ] potentially running migration '20220317-align_num_heartbeats'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810502975Z [INFO ] no need to migrate '20220317-align_num_heartbeats'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810578747Z [INFO ] potentially running migration '20220318-mysql_timestamp_precision'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810721077Z [INFO ] no need to migrate '20220318-mysql_timestamp_precision'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.81079248Z [INFO ] potentially running migration '202203191-drop_diagnostics_user'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.810937691Z [INFO ] no need to migrate '202203191-drop_diagnostics_user'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.811025183Z [INFO ] potentially running migration '20221016-drop_rank_column'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.811170988Z [INFO ] no need to migrate '20221016-drop_rank_column'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.81124974Z [INFO ] potentially running migration '20230219-add_astro_language'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.811391188Z [INFO ] no need to migrate '20230219-add_astro_language'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.811529028Z [INFO ] potentially running migration '20230219-add_subscription_renewal'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.81168032Z [INFO ] no need to migrate '20230219-add_subscription_renewal'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.811757318Z [INFO ] potentially running migration '20231023-fill_last_branch'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.811899685Z [INFO ] no need to migrate '20231023-fill_last_branch'
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.813408266Z [INFO ] creating job queue 'wakapi.default' (1 workers)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.813537412Z [INFO ] creating job queue 'wakapi.processing' (1 workers)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.813648328Z [INFO ] creating job queue 'wakapi.reports' (1 workers)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.813718658Z [INFO ] creating job queue 'wakapi.housekeeping' (1 workers)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.813804306Z [INFO ] creating job queue 'wakapi.mail' (1 workers)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.814018565Z [INFO ] creating job queue 'wakapi.imports' (1 workers)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.814088283Z [INFO ] scheduling report generation
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.814169848Z [INFO ] scheduling summary aggregation
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.814212828Z [INFO ] scheduling inactive users cleanup
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.815166512Z [INFO ] scheduling project stats cache pre-warming
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.815206486Z [INFO ] scheduling total time counting
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.815232136Z [INFO ] scheduling first data computing
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.817157959Z [INFO ] computing users' first data
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.817275989Z [INFO ] scheduling leaderboard generation
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.821486823Z [INFO ] 👉 Listening for HTTPS on 0.0.0.0:3080... ✅
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.821608589Z [INFO ] 👉 Listening for HTTPS on [::1]:3080... ✅
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: 2024-06-19T00:34:13.823527563Z [ERROR] listen tcp [::1]:3080: bind: address already in use
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: panic: listen tcp [::1]:3080: bind: address already in use
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: goroutine 57 [running]:
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: github.com/emvi/logbuch.(*StandardFormatter).Pnc(0x25423c0?, {0xc000ed86c0?, 0xc000ed86c0?}, {0x0?, 0x0?, 0x0?})
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]:         /home/ec2-user/go/pkg/mod/github.com/emvi/logbuch@v1.2.0/standard_formatter.go:57 +0x37
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: github.com/emvi/logbuch.(*Logger).Fatal(0x25423c0, {0xc000ed86c0, 0x33}, {0x0, 0x0, 0x0})
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]:         /home/ec2-user/go/pkg/mod/github.com/emvi/logbuch@v1.2.0/logger.go:132 +0x72
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: github.com/emvi/logbuch.Fatal(...)
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]:         /home/ec2-user/go/pkg/mod/github.com/emvi/logbuch@v1.2.0/funcs.go:55
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: main.listen.func2()
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]:         /home/ec2-user/go/pkg/mod/github.com/muety/wakapi@v0.0.0-20240613185946-af85a65d0cb7/main.go:394 +0x6a
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: created by main.listen in goroutine 1
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]:         /home/ec2-user/go/pkg/mod/github.com/muety/wakapi@v0.0.0-20240613185946-af85a65d0cb7/main.go:392 +0x5e5
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal systemd[1]: wakapi.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal systemd[1]: wakapi.service: Failed with result 'exit-code'.
```

 i have not yet been able to access the website from any public url , while on port forwarding i can see that it works when the listen_ipv4 is set to localhost in the config.yml.

**Comment by muety** [comment]
If you set the listen IP to localhost, you'll also only be able to access it from localhost. Set it to `0.0.0.0` (or the specific IP of your public network interface) to expose it publicly instead. Also the `bind: address already in use` suggests that the IPv6 port 3080 is already blocked. Perhaps you have another Wakapi instance already running on that port? You can check with `netstat -tulpn | grep 3080`.

**Reply by palash-sarate** [reply]
I am already using 0.0.0.0 for ipv4 and ::1 for ipv6 as can be verified in logs and the config file. 

I get error
```
[ERROR] listen tcp [::1]:3080: bind: address already in use
Jun 19 00:34:13 ip-172-31-44-5.ap-south-1.compute.internal wakapi[3765]: panic: listen tcp [::1]:3080: bind: address already in use.
```
even when I, stop/delete all wakapi services, reboot the instance and then again try to start the service. This is happening even if I change the port and restart the service I get same error for the new port.

netstat -tulpn | grep {port}
this return nothing even when run with sudo, (replacing the {port} with the port currently set in the config, which gave the above error) 

Thanks for you time.

**Reply by YC** [reply]
This is reproducible with:
```
listen_ipv4: 0.0.0.0              # set to '-' to disable ipv4
listen_ipv6: ::1                    # set to '-' to disable ipv6
```

Perhaps try
```
listen_ipv4: '-'              # set to '-' to disable ipv4
listen_ipv6: ::0                    # set to '-' to disable ipv6
```

**Reply by muety** [reply]
I was also gonna suggest to try with wither v4 or v6 only to see if that changes anything. Also, the config you posted looks somewhat weird, as all colons (`:`) seem to be missing? :thinking: 

Moreover, did you perhaps have a firewall configured on AWS in front of your EC2 instance?

**Reply by palash-sarate** [reply]
if you remove the .txt extension at last it should view properly as  yaml. Uploading  .yml was prohibited that's why uploaded it as txt. no firewall is configured on ec2, But do u think it will interfere with setting the listen address in wakapi go app?

**Reply by palash-sarate** [reply]
> This is reproducible with:
> 
> ```
> listen_ipv4: 0.0.0.0              # set to '-' to disable ipv4
> listen_ipv6: ::1                    # set to '-' to disable ipv6
> ```
> 
> Perhaps try
> 
> ```
> listen_ipv4: '-'              # set to '-' to disable ipv4
> listen_ipv6: ::0                    # set to '-' to disable ipv6
> ```

i dont get that error now. 👉 Listening for HTTPS on [::0]:3080... ✅ the service starts successfully. But how will we be able to access over public url if ipv4 is not. sorry if its a dumb question.

**Reply by muety** [reply]
https://[1234:5678:...]::3080

**Reply by YC** [reply]
From my brief testing, the application should be listening on both IPv4 and IPv6 with that configuration. 
If you only want IPv4, then maybe:

```
listen_ipv4: 0.0.0.0              # set to '-' to disable ipv4
listen_ipv6: '-'                    # set to '-' to disable ipv6
```

**Reply by palash-sarate** [reply]
> From my brief testing, the application should be listening on both IPv4 and IPv6 with that configuration. If you only want IPv4, then maybe:
> 
> ```
> listen_ipv4: 0.0.0.0              # set to '-' to disable ipv4
> listen_ipv6: '-'                    # set to '-' to disable ipv6
> ```

I tried this and the app starts listening correctly to 0.0.0.0:3000, but it didn't serve the app on EC2 instance public URL. I have allowed all https, http traffic using the security group so that wont be a problem. are there any next steps to this. Is there any way to verify the page is being served using CURL on the ec2 instance? 

Thank you for your time

**Reply by muety** [reply]
> I have allowed all https, http traffic

You'll specifically have to enable incoming TCP for the port that your app is listening on.

**Reply by YC** [reply]
> Is there any way to verify the page is being served using CURL on the ec2 instance?

```
$ curl http://localhost:3000
```

I wouldn't recommend exposing a service over the Internet over http though. 
You should use https or set up a reverse proxy with caddy/nginx/haproxy etc. to serve multiple apps over 80/443 (which you presumably allowed in the security group).

**Reply by palash-sarate** [reply]
> > Is there any way to verify the page is being served using CURL on the ec2 instance?
> 
> ```
> $ curl http://localhost:3000
> ```
> 
> I wouldn't recommend exposing a service over the Internet over http though. You should use https or set up a reverse proxy with caddy/nginx/haproxy etc. to serve multiple apps over 80/443 (which you presumably allowed in the security group).

as application is running fine on local host it return the landing page correctly. the dns to public url resolves correctly but when trying to curl the public URL connection fails 
```
[ec2-user@ip-172-31-44-5 ~]$ curl https://wakatime.zeuron.ai
curl: (7) Failed to connect to wakatime.zeuron.ai port 443 after 55 ms: Couldn't connect to server
```

I am now trying to setup nginx as reverse proxy that also handle ssl termination , as this did not work.

**Reply by muety** [reply]
If you're hosting nothing but Wakapi on the server, setting up a reverse proxy is probably overkill tbh. You can have Wakapi serve via TLS itself using the `server.tls_*` config variables. Also, for initial testing, you can also just go with plain HTTP as long as you switch to HTTPS in the long run.

Also, again, if Wakapi is listening on `3000`, you'll need a firewall rule to allow incoming traffic to that port and then access https://wakatime.zeuron.ai:3000.

**Reply by palash-sarate** [reply]
> If you're hosting nothing but Wakapi on the server, setting up a reverse proxy is probably overkill tbh. You can have Wakapi serve via TLS itself using the `server.tls_*` config variables. Also, for initial testing, you can also just go with plain HTTP as long as you switch to HTTPS in the long run.
> 
> Also, again, if Wakapi is listening on `3000`, you'll need a firewall rule to allow incoming traffic to that port and then access https://wakatime.zeuron.ai:3000.

I think this would work, is there any way to avoid any port number. I tried setting it to 443, 80 but got permission denied.

**Reply by muety** [reply]
The `CapabilityBoundingSet=CAP_NET_BIND_SERVICE` setting in the service file should account for that. Perhaps google for "linux allow bind privileged ports".

**Comment by palash-sarate** [comment]
Hi, I am trying to use helm chart now on my eks cluster. but I am getting an error in pod logs
```
2024-06-19T15:15:37.309231072Z [ERROR] unexpected end of JSON input
panic: unexpected end of JSON input
goroutine 1 [running]:
github.com/emvi/logbuch.(*StandardFormatter).Pnc(0x2520100?, {0x11d5162?, 0x11d5162?}, {0x0?, 0x0?, 0x0?})
/go/pkg/mod/github.com/emvi/logbuch@v1.2.0/standard_formatter.go:57 +0x37
github.com/emvi/logbuch.(*Logger).Fatal(0x2520100, {0x11d5162, 0x1c}, {0x0, 0x0, 0x0})
/go/pkg/mod/github.com/emvi/logbuch@v1.2.0/logger.go:132 +0x72
github.com/emvi/logbuch.Fatal(...)
/go/pkg/mod/github.com/emvi/logbuch@v1.2.0/funcs.go:55
github.com/muety/wakapi/config.readColors()
/src/config/config.go:426 +0x145
github.com/muety/wakapi/config.Load({0x11c09f5, 0xa}, {0x11bcc34, 0x7})
/src/config/config.go:469 +0x2ad
main.main()
/src/main.go:122 +0x138
```

I have used the helm chart from suggested repo as is since i didnt see any need to change anything except smtp settings, and simply used helm to install. 

I will create a guide to host on eks or ec2 if I succeed in either.
